frb_codegen/src/utils/misc.rs

```diff
@@ -1,25 +1,100 @@
+use std::cell::RefCell;
 use std::collections::HashSet;
 
 use std::fmt::Display;
 use std::fs;
-use std::hash::Hash;
+
 use std::path::Path;
 
 use crate::commands::cargo_expand;
+use crate::ir::{IrType, IrTypeTrait};
+
 use crate::utils::consts::DART_KEYWORDS;
 use anyhow::anyhow;
 use convert_case::{Case, Casing};
 use pathdiff::diff_paths;
 
-pub fn mod_from_rust_path(code_path: &str, crate_path: &str) -> String {
-    Path::new(code_path)
-        .strip_prefix(Path::new(crate_path).join("src"))
-        .unwrap()
-        .with_extension("")
-        .into_os_string()
-        .into_string()
-        .unwrap()
-        .replace('/', "::")
+/// get the raw or shared API module name
+pub fn mod_from_rust_path(config: &crate::Opts, get_shared_mod: bool) -> Option<String> {
+    fn get_module_name(code_path: &str, crate_path: &str) -> String {
+        Path::new(code_path)
+            .strip_prefix(Path::new(crate_path).join("src"))
+            .unwrap()
+            .with_extension("")
+            .into_os_string()
+            .into_string()
+            .unwrap()
+            .replace('/', "::")
+    }
+
+    let output = match !config.shared {
+        true => {
+            if !get_shared_mod {
+                get_module_name(&config.rust_input_path, &config.rust_crate_dir)
+            } else {
+                // get the shared module in the regular block
+                if is_multi_blocks_case(None) {
+                    get_module_name(
+                        &config.shared_rust_output_path.clone().unwrap(),
+                        &config.rust_crate_dir,
+                    )
+                } else {
+                    "".into()
+                }
+            }
+        }
+        false => {
+            // Whatever `get_shared_mod` is, return the shared module name for a shared block
+            get_module_name(
+                &config.shared_rust_output_path.clone().unwrap(),
+                &config.rust_crate_dir,
+            )
+        }
+    };
+
+    if output.is_empty() {
+        None
+    } else {
+        Some(output)
+    }
+}
+
+thread_local!(pub static IS_MULTI_BLOCKS_CASE: RefCell<Option<bool>> = RefCell::new(None));
+
+pub fn is_multi_blocks_case(all_configs: Option<&[crate::Opts]>) -> bool {
```

Comment:
well.. I am really not sure whether this is ok or not. This seems to be essentially global variables, which we know is very bad and can cause a ton of troubles.

The thing most similar to such a global variable is something like a big "context" that is passed around everywhere. Say,

```
struct Context {
  is_multi_blocks_case: bool
}
```

and the `context: &Context` is an arg passed everywhere.

But that may not also be best depending on the scenario. Another way may be, accurately name it, say `struct ConfigThatGeneratorNeeds { ... }`.